migrations/repotracker_cleanup.go

```diff
@@ -0,0 +1,452 @@
+package migrations
+
+import (
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/build"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/evergreen-ci/evergreen/model/version"
+	"github.com/mongodb/anser"
+	"github.com/mongodb/anser/db"
+	"github.com/mongodb/anser/model"
+	"github.com/mongodb/grip"
+	"github.com/pkg/errors"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	buildCollection   = "builds"
+	versionCollection = "versions"
+	taskCollection    = "tasks"
+
+	requesterKey = "r"
+	statusKey    = "status"
+)
+
+func orphanedVersionCleanupGenerator(env anser.Environment, db string, limit int) (anser.Generator, error) {
+	const migrationName = "clean_orphaned_versions"
+
+	if err := env.RegisterManualMigrationOperation(migrationName, makeOrphanedVersionCleanup(db)); err != nil {
+		return nil, err
+	}
+
+	opts := model.GeneratorOptions{
+		NS: model.Namespace{
+			DB:         db,
+			Collection: versionCollection,
+		},
+		Query: bson.M{
+			requesterKey: evergreen.RepotrackerVersionRequester,
+			statusKey:    evergreen.VersionCreated, // TODO is this true?
+		},
+		Limit: limit,
+		JobID: "migration-versions-orphans",
+	}
+
+	return anser.NewManualMigrationGenerator(env, opts, migrationName), nil
+}
+
+func makeOrphanedVersionCleanup(database string) db.MigrationOperation {
+	const (
+		idKey            = "_id"
+		buildStatusesKey = "build_variants_status"
+		buildIDsKey      = "builds"
+	)
+	return func(session db.Session, rawD bson.RawD) error {
+		defer session.Close()
+
+		bstatuses := []version.BuildStatus{}
+		builds := []string{}
+		versionID := ""
+
+		for i := range rawD {
+			switch rawD[i].Name {
+			case idKey:
+				if err := rawD[i].Value.Unmarshal(&versionID); err != nil {
+					return errors.Wrap(err, "error unmarshaling id")
+				}
+
+			case buildStatusesKey:
+				if err := rawD[i].Value.Unmarshal(&bstatuses); err != nil {
+					return errors.Wrap(err, "error unmarshaling build statuses")
+				}
+
+			case buildIDsKey:
+				if err := rawD[i].Value.Unmarshal(&builds); err != nil {
+					return errors.Wrap(err, "error unmarshaling builds")
+				}
+			}
+		}
+
+		buildMap := map[string]bool{}
+
+		newBuilds := []string{}
+		for _, buildID := range builds {
+			_, ok := buildMap[buildID]
+			if !ok {
+				query := session.DB(database).C(buildCollection).FindId(buildID)
+				err := query.One(nil)
+				if err == mgo.ErrNotFound {
+					continue
+				}
+				if err != nil {
+					return errors.Wrapf(err, "error fetching build %s", buildID)
+				}
+				buildMap[buildID] = true
+			}
+
+			newBuilds = append(newBuilds, buildID)
+		}
+
+		newBuildStatuses := []version.BuildStatus{}
+		for _, bstatus := range bstatuses {
+			_, ok := buildMap[bstatus.BuildId]
+			if !ok {
+				query := session.DB(database).C(buildCollection).FindId(bstatus.BuildId)
+				err := query.One(nil)
```

Comment:
does it make more sense to pass a query to count rather than do this findOne into a nil?